src/pkg/project/dao/dao.go

```diff
@@ -0,0 +1,178 @@
+	// GetByName get project instance by name
+	GetByName(ctx context.Context, name string) (*models.Project, error)
+	// List list projects
+	List(ctx context.Context, query *q.Query) ([]*models.Project, error)
+}
+
+// New returns an instance of the default DAO
+func New() DAO {
+	return &dao{}
+}
+
+type dao struct{}
+
+// Create create a project instance
+func (d *dao) Create(ctx context.Context, project *models.Project) (int64, error) {
+	if project.OwnerID <= 0 {
+		return 0, errors.BadRequestError(nil).WithMessage("owner_id invalid")
+	}
+
+	var projectID int64
+
+	h := func(ctx context.Context) error {
+		o, err := orm.FromContext(ctx)
+		if err != nil {
+			return err
+		}
+
+		now := time.Now()
+		project.CreationTime = now
+		project.UpdateTime = now
+
+		projectID, err = o.Insert(project)
+		if err != nil {
+			return err
+		}
+
+		member := &Member{
+			ProjectID:    projectID,
+			EntityID:     project.OwnerID,
+			Role:         common.RoleProjectAdmin,
+			EntityType:   common.UserMember,
+			CreationTime: now,
+			UpdateTime:   now,
+		}
+
+		if _, err := o.Insert(member); err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	if err := orm.WithTransaction(h)(ctx); err != nil {
+		return 0, err
+	}
+
+	return projectID, nil
+}
+
+// Delete delete the project instance by id
+func (d *dao) Delete(ctx context.Context, id int64) error {
+	project, err := d.Get(ctx, id)
+	if err != nil {
+		return err
+	}
+
+	project.Deleted = true
+	project.Name = fmt.Sprintf("%s#%d", project.Name, project.ProjectID)
+
+	o, err := orm.FromContext(ctx)
+	if err != nil {
+		return err
+	}
+
+	_, err = o.Update(project, "deleted", "name")
+	return err
+}
+
+// Get get project instance by id
+func (d *dao) Get(ctx context.Context, id int64) (*models.Project, error) {
+	o, err := orm.FromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	sql := `select p.project_id, p.name, p.registry_id, u.username as owner_name, p.owner_id, p.creation_time, p.update_time
+	from project p left join harbor_user u on p.owner_id = u.user_id where p.deleted = false and p.project_id = ?`
+
+	projects := []*models.Project{}
+	count, err := o.Raw(sql, id).QueryRows(&projects)
+	if err != nil {
+		return nil, err
+	}
+
+	if count == 0 {
+		return nil, errors.NotFoundError(nil).WithMessage("project %d not found", id)
+	}
+
+	return projects[0], nil
+}
+
+// GetByName get project instance by name
+func (d *dao) GetByName(ctx context.Context, name string) (*models.Project, error) {
+	o, err := orm.FromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	sql := `select p.project_id, p.name, p.registry_id, u.username as owner_name, p.owner_id, p.creation_time, p.update_time
+	from project p left join harbor_user u on p.owner_id = u.user_id where p.deleted = false and p.name = ?`
+
+	projects := []*models.Project{}
+	count, err := o.Raw(sql, name).QueryRows(&projects)
+	if err != nil {
+		return nil, err
+	}
+
+	if count == 0 {
```

Comment:
in the previous version, the `GetByName` doesn't raise NotFound error, please be care for any break change.

Reply:
In the beginning, the `Get` and `GetByName` methods of project controller will return `NotFound` error when the project was not found so it's ok for the DAO to return `NotFound` error here.